/*
* @license Apache-2.0
*
* Copyright (c) 2021 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

/* eslint-disable @typescript-eslint/no-unused-expressions */

import CircularBuffer = require( './index' );


// FIXME: add tests


// TESTS //

// The function returns a circular buffer instance...
{
	new CircularBuffer( 3 ); // $ExpectType CircularBuffer<unknown>
}

// The compiler throws an error if the constructor function is provided an invalid number of arguments...
{
	new CircularBuffer(); // $ExpectError
	new CircularBuffer( 3, 4 ); // $ExpectError
}

// Test the behavior when initializing with a capacity of zero
{
    const buffer = new CircularBuffer(0);
    buffer.push(1); // $ExpectError
    buffer.pop(); // $ExpectError
    buffer.length; // $ExpectType number
}

// Test the behavior when pushing elements of different types
{
    const buffer = new CircularBuffer<string>(3);
    buffer.push('a');
    buffer.push('b');
    buffer.push(3); // $ExpectError
    buffer.push(null); // $ExpectError
}

// Test the behavior when popping from an empty buffer
{
    const buffer = new CircularBuffer<number>(3);
    buffer.pop(); // $ExpectError
}

// Test the behavior when clearing the buffer
{
    const buffer = new CircularBuffer<number>(3);
    buffer.push(1);
    buffer.push(2);
    buffer.clear();
    buffer.length; // $ExpectType number
    buffer.pop(); // $ExpectError
}
